Hash-Tables-and-Hashing Functions.md

```diff
@@ -0,0 +1,167 @@
+   5                 |    Spain         |    Madrid
+   6                 |    France        |    Paris
+   7                 |    England       |    London
+   8                 |                  | 
+   9                 |                  |
+   10                |                  |
+   11                |  Switzerland     |    Berne
+
+
+Now, in this specific example things work quite well. 
+Our array needs to be big enough to accommodate the longest string, but in this case that’s only 11 slots. 
+And we do waste a bit of space because, for example, there are no 1-letter keys in our data, nor keys between 8 and 10 letters. But in this case, the waste isn’t so bad either. And taking the length of a string is nice and fast, so is the process of finding the value associated with a given key (certainly faster than doing up to five string comparisons).
+
+But, what do we do if our dataset has a string which has more than 11 characters? 
+What if we have one another word with 5 characters, "India", and try assigning it to an index using our hash function. Since the index 5 is already occupied, we have to make a call on what to do with it. This is called a collision. 
+If our dataset had a string with thousand characters, and you make an array of thousand indices to store the data, it would result in a wastage of space. If our keys were random words from English, where there are so many words with same length, using length as a hashing function would be fairly useless.
+
+
+## Collision Handling
+
+Two basic methods are used to handle collisions.
+   1. Separate Chaining
+   2. Open Addressing
+
+### Separate Chaining
+
+Hash collision handling by separate chaining, uses an additional data structure, preferrably linked list for dynamic allocation, into buckets. In our example, when we add India to the dataset, it is appended to the linked list stored at the index 5, then our table would look like this.
+
+Position             | Linked List Heads                  | 
+---------------------|------------------------------------|
+   1                 |                                    |
+   2                 |                                    |
+   3                 |                                    |
+   4                 |    [Cuba-Havana]                   |    
+   5                 |    [Spain-Madrid] -> [India-Delhi] |
+   6                 |    [France-Paris]                  |
+   7                 |    [England-London]                |
+   8                 |                                    | 
+   9                 |                                    |
+   10                |                                    |
+   11                |    [Switzerland-Berne]             |
+
+
+So to find an item we first go to the bucket then compare keys. This is a popular method, and if link list is used the hash never fills up. The cost for `get(k)` is on average `O(n)` where n is the number of keys in the bucket, total number of keys be N. 
```

Comment:
~~So to find an item~~ To find an item, 

~~then compare keys.~~ and then we compare keys.

~~if link list~~ if a list of links